blog/2024-12-29-rust-gpu-0.10.md

```diff
@@ -0,0 +1,250 @@
+patch out others](https://github.com/Rust-GPU/rust-gpu/pull/170), unblocking the update
+and buying us time to [figure out the best way
+forward](https://github.com/Rust-GPU/rust-gpu/issues/182).
+
+Rust GPU is an "out-of-tree" compiler backend. This enables fast iteration while we
+explore the problem space, but also makes us more susceptible to this sort of breakage.
+One of our long-term goals is to be a supported `rustc` backend.
+
+### **Mesh shader support**
+
+Rust GPU maintainer [@Firestar99](https://github.com/firestar99) and contributor
+[@BeastLe9enD](https://github.com/BeastLe9enD) added support for [mesh
+shaders](https://github.com/Rust-GPU/rust-gpu/pull/44).
+
+[Mesh shaders](https://www.khronos.org/blog/mesh-shading-for-vulkan) are a modern GPU
+feature that replace traditional vertex and geometry shaders. They give GPU developers
+more control over the entire pipeline by letting them directly process batches of
+vertices and primitives in compute-like shaders. This allows for more efficient culling,
+level-of-detail calculations, and custom pipeline logic—all on the GPU.
+
+An example of a Rust GPU mesh shader that outputs points:
+
+```
+use spirv_std::arch::set_mesh_outputs_ext;
+use spirv_std::glam::{UVec2, Vec4};
+use spirv_std::spirv;
+
+#[spirv(mesh_ext(
+    threads(1),
+    output_vertices = 1,
+    output_primitives_ext = 1,
+    output_points
+))]
+pub fn main(
+    #[spirv(position)] positions: &mut [Vec4; 1],
+    #[spirv(primitive_point_indices_ext)] indices: &mut [u32; 1],
+) {
+    unsafe {
+        set_mesh_outputs_ext(1, 1);
+    }
+
+    positions[0] = Vec4::new(-0.5, 0.5, 0.0, 1.0);
+
+    indices[0] = 0;
+}
+```
+
+An example of a Rust GPU mesh task shader:
+
+```
+use spirv_std::arch::emit_mesh_tasks_ext;
+use spirv_std::spirv;
+
+#[spirv(task_ext(threads(1)))]
+pub fn main() {
+    unsafe {
+        emit_mesh_tasks_ext(1, 2, 3);
+    }
+}
+```
+
+### **Subgroup support**
+
+[@Firestar99](https://github.com/firestar99) also added support for subgroups via
+[subgroup intrinsics](https://github.com/Rust-GPU/rust-gpu/pull/14).
+
+[Subgroups](https://www.khronos.org/blog/vulkan-subgroup-tutorial) are small groups of
+threads within a workgroup that can share data and perform synchronized operations more
+efficiently. For example, using subgroup intrinsics you can:
+
+- Perform reductions (e.g., sum, min, max) across threads in a subgroup.
+- Share intermediate results without relying on global memory, reducing latency.
+- Implement algorithms like prefix sums or parallel sorting more effectively.
+
+Here is a simple Rust GPU example to demonstrate subgroup reduction:
+
+```rust
+use glam::UVec3;
+use spirv_std::spirv;
+
+unsafe fn subgroup_i_add_reduce(value: u32) -> u32 {
+    spirv_std::arch::subgroup_i_add(value)
+}
+
+#[spirv(compute(threads(32, 1, 1)))]
+pub fn main(#[spirv(local_invocation_id)] local_invocation_id: UVec3) {
+    unsafe {
+        subgroup_i_add_reduce(local_invocation_id.x);
+    }
+}
+```
```

Comment:
I don't like this example at all, the tests for subgroups were just to check the asm, not to be an actual practical example. 

Here's a practical example from my codebase: https://gitlab.com/spacegamedev/space-rust/-/blob/meshlet_piping/space-engine-shader/src/renderer/allocation_buffer.rs?ref_type=heads#L34

I may make you a good example over the next few days, if I find the time that is.

Reply:
I just removed the example. If you get time, just add one back. No worries if not!